File: lc_soln/0110_balanced_binary_tree/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution
{
    public bool IsBalanced(TreeNode root)
    {
        return Dfs(root).isBalanced;
    }

    (bool isBalanced, int height) Dfs(TreeNode root)
    {
        if (root is null)
        {
            return (true, 0);
        }
        (bool leftBalanced, int leftHeight) = Dfs(root.left);
        (bool rightBalanced, int rightHeight) = Dfs(root.right);
        return (
            leftBalanced && rightBalanced && Math.Abs(leftHeight - rightHeight) < 2,
            1 + Math.Max(leftHeight, rightHeight)
        );
    }
}

File: lc_soln/0110_balanced_binary_tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int height;
        return dfs(root, height);
    }

    bool dfs(TreeNode* root, int& height) {
        if (root == nullptr) {
            height = 0;
            return true;
        }
        int left_height;
        bool left = dfs(root->left, left_height);
        int right_height;
        bool right = dfs(root->right, right_height);
        height = 1 + max(left_height, right_height);
        return left && right && std::abs(left_height - right_height) < 2;
    }
};

int main() {
    return 0;
}
